coderd/rbac/policy.rego

```diff
@@ -19,121 +19,98 @@ bool_flip(b) = flipped {
     flipped = true
 }
 
-# perms_grant returns a set of boolean values {true, false}.
-# True means a positive permission in the set, false is a negative permission.
-# It will only return `bool_flip(perm.negate)` for permissions that affect a given
-# resource_type, and action.
-# The empty set is returned if no relevant permissions are found.
-perms_grant(permissions) = grants {
-    # If there are no permissions, this value is the empty set {}.
-    grants := { x |
-        # All permissions ...
-        perm := permissions[_]
-        # Such that the permission action, and type matches
-        perm.action in [input.action, "*"]
-        perm.resource_type in [input.object.type, "*"]
-        x := bool_flip(perm.negate)
-    }
+# number is a quick way to get a set of {true, false} and convert it to
+#  -1: {false, true} or {false}
+#   0: {}
+#   1: {true}
+number(set) = c {
+	count(set) == 0
+    c := 0
 }
 
-# Site & User are both very simple. We default both to the empty set '{}'. If no permissions are present, then the
-# result is the default value.
-default site = {}
-site = grant {
-    # Boolean set for all site wide permissions.
-    grant = { v | # Use set comprehension to remove duplicate values
-        # For each role, grab the site permission.
-        # Find the grants on this permission list.
-        v = perms_grant(input.subject.roles[_].site)[_]
-    }
+number(set) = c {
+	false in set
+    c := -1
 }
 
-default user = {}
-user = grant {
-    # Only apply user permissions if the user owns the resource
-    input.object.owner != ""
-    input.object.owner == input.subject.id
-    grant = { v |
-        # For each role, grab the user permissions.
-        # Find the grants on this permission list.
-        v = perms_grant(input.subject.roles[_].user)[_]
-    }
+number(set) = c {
+	not false in set
+	set[_]
+    c := 1
 }
 
-# Organizations are more complex. If the user has no roles that specifically indicate the org_id of the object,
-# then we want to block the action. This is because that means the user is not a member of the org.
-# A non-member cannot access any org resources.
 
-# org_member returns the set of permissions associated with a user if the user is a member of the
-# organization
-org_member = grant {
-    input.object.org_owner != ""
-    grant = { v |
-        v = perms_grant(input.subject.roles[_].org[input.object.org_owner])[_]
+default site = 0
```

Comment:
before this policy had some good commentary about how it worked, but we've lost that in the new version.

I'd also like some commentary that explains why the policy is organized this way, rather than the way we had it before.  Presumably it's because it helps with partial eval, but it's not clear to me from reading this rego why it would be better or worse for partials.

Reply:
Added a lot of commentary